notebooks/time_series/apple_stock_prices.livemd

```diff
@@ -0,0 +1,264 @@
+
+An architecture known as Long Term Short Memory (LSTM) will be used for the RNN.
+
+<!-- livebook:{"break_markdown":true} -->
+
+First, we load the historical data. Subsequently, preprocessing will be conducted to utilize the information with an RNN model. The time series data is already normalized. This helps mitigate significant numerical issues associated with how activation functions like tanh transform very large numbers (whether positive or negative), and it aids in avoiding problems with calculating derivatives.
+
+```elixir
+raw_csv =
+  "https://raw.githubusercontent.com/santiago-imelio/datasets/main/normalized_apple_prices.csv"
+  |> Req.get!()
+  |> Map.get(:body)
+
+data =
+  Regex.split(~r/\r|\n|\r\n/, raw_csv)
+  |> List.pop_at(-1)
+  |> elem(1)
+  |> Enum.map(&Float.parse/1)
+  |> Enum.map(&elem(&1, 0))
+  |> Nx.tensor()
+```
+
+The series is normalized to belong within the range [-1, 1] using min-max scaling. It's also common to see applications where normalization is performed using standard deviation. Let's visualize the loaded data.
+
+```elixir
+{series_size} = Nx.shape(data)
+days = Nx.iota({series_size})
+
+Tucan.lineplot([Prices: data, Days: days], "Days", "Prices")
+|> Tucan.set_size(600, 400)
+```
+
+## Sliding window
+
+Generally, a time-series is mathematically represented as:
+
+$$
+\langle s_0, s_1, s_2, \ldots, s_P \rangle
+$$
+
+where $s_p$ is the numerical value of the series at time interval $p$, and $P$ is the total length of the series. To apply an RNN, the prediction should be treated as a regression problem. This involves using a sliding window to construct a set of input-output pairs on which regression will be applied.
+
+For example, with a window size $T = 3$, the following pairs need to be produced:
+
+$$
+\begin{array}{c|c}
+  \text{Input} & \text{Output}\\
+  \hline \langle s_{1},s_{2},s_{3}\rangle & s_{4} \\
+  \langle s_{2},s_{3},s_{4} \rangle & s_{5}  \\
+  \vdots & \vdots \\
+  \langle s_{P-3},s_{P-2},s_{P-1} \rangle & s_{P}
+\end{array}
+$$
+
+In each pair, the input consists of a sequence of $T$ consecutive values from the series, and the output is the subsequent value immediately following the input sequence. This approach prepares the data for training the RNN model for time-series prediction.
+
+```elixir
+sliding_window = fn series, win_size ->
+  {p} = Nx.shape(series)
+
+  x_idx =
+    series
+    |> Nx.shape()
+    |> Nx.iota()
+    |> Nx.reshape({:auto, 1})
+    |> Nx.vectorize(:elements)
+
+  y_idx =
+    series
+    |> Nx.shape()
+    |> Nx.iota()
+    |> Nx.add(win_size)
+    |> Nx.reshape({:auto, 1})
+    |> Nx.slice([0, 0], [p - win_size, 1])
+
+  x =
+    Nx.slice(series, [x_idx[0]], [win_size])
+    |> Nx.devectorize(keep_names: false)
+    |> Nx.slice([0, 0], [p - win_size, win_size])
+
+  y = Nx.gather(series, y_idx, axes: [0])
+
+  {x, y}
+end
+```
+
+Next we test our sliding window function with first 10 numbers of the Fibonacci sequence.
```

Comment:
```suggestion
Next, we test our sliding window function with first 10 numbers of the Fibonacci sequence.
```

Reply:
It might be good to explain why or how using the fibonacci sequence here is a good "test" and what specifically you're looking for out of this function